### PR TITLE
Add zero() constructor for Quantities

### PR DESCRIPTION
**Pull request by llucax**
It is a very common case which becomes very verbose, and it includes unit information that is completely irrelevant (zero is zero in any unit).

### CODE REVIEW

File: tests/timeseries/test_quantities.py [hunk 1]
```diff
@@ -42,6 +42,56 @@ class Fz2(
...
+    assert Frequency.zero().as_hertz() == 0.0
+    assert Frequency.zero().as_megahertz() == 0.0
+    assert Frequency.zero() is Frequency.zero()  # It is a "singleton"
+
+    assert Percentage.from_percent(0.0) == Percentage.zero()
+    assert Percentage.from_fraction(0.0) == Percentage.zero()
+    assert isinstance(Percentage.zero(), Percentage)
+    assert Percentage.zero().as_percent() == 0.0
+    assert Percentage.zero().as_fraction() == 0.0
+    assert Percentage.zero() is Percentage.zero()  # It is a "singleton"
```
Comment (Marenz) [inline tests/timeseries/test_quantities.py:98]:
```py
x = Percentage.zero()
x = x + Percentage.from_percent(1)
assert Percentage.zero() == Percentage.from_percent(0)
```

would this assert fail?

File: src/frequenz/sdk/timeseries/_quantities.py [hunk 2]
```diff
@@ -62,6 +62,30 @@ def __init_subclass__(cls, exponent_unit_map: dict[int, str]) -> None:
...
+    """Cache for zero singletons.
+
+    This is a workaround for mypy getting confused when using @functools.cache and
+    @classmethod combined with returning Self. It believes the resulting type of this
+    method is Self and complains that members of the actual class don't exist in Self,
+    so we need to implement the cache ourselves.
+    """
+
+    @classmethod
+    def zero(cls) -> Self:
```
Comment (shsms) [inline src/frequenz/sdk/timeseries/_quantities.py:78]: I guess class methods can't be properties?

File: tests/timeseries/test_quantities.py (hunk 1 above)
Comment (shsms) [reply tests/timeseries/test_quantities.py:98]: I think it wouldn't,  because `x + ...` generates a new object,  and that gets bound to `x`,  and the zero object's reference count would just get decremented.  It is not possible to get the deref behaviour of other languages in Python I think (at least easily).

Comment (Marenz) [reply tests/timeseries/test_quantities.py:98]:
```py
        x = Percentage.zero()
        x += Percentage.from_percent(1.0)
>       assert Percentage.zero() == Percentage.from_percent(0.0)
E       AssertionError: assert Percentage(value=1.0, exponent=0) == Percentage(value=0.0, exponent=0)
```

Comment (Marenz) [reply tests/timeseries/test_quantities.py:98]: but true, with `x = x + ..` it doesn't fail

File: src/frequenz/sdk/timeseries/_quantities.py (hunk 2 above)
Comment (Marenz) [reply src/frequenz/sdk/timeseries/_quantities.py:78]: No, I think it works

File: tests/timeseries/test_quantities.py (hunk 1 above)
Comment (shsms) [reply tests/timeseries/test_quantities.py:98]: :o I forgot we implemented +=, etc.  So now we have to decide whether to keep += or not cache zeros?  My feeling is that zero values are rarely created as part of the hot-path,  so we can just create a new object and not cache them.

Comment (Marenz) [reply tests/timeseries/test_quantities.py:98]: You could do something terrible and check in `+=` if you are a singleton and reset the according cache

Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: You can't change the object returned by `Percentage.zero()` unless you do this: `Percentage._zero_cache[Percentage] = x`. Variables in Python are bound to objects, you are just rebinding `x` when doing `x = ...`, nothing more.

File: src/frequenz/sdk/timeseries/_quantities.py (hunk 2 above)
Comment (llucax) [reply src/frequenz/sdk/timeseries/_quantities.py:78]: Interesting, it occurred to me but I didn't check, I can try.

Comment (llucax) [reply src/frequenz/sdk/timeseries/_quantities.py:78]: Python (tests) works, but `pylint` doesn't :roll_eyes: 
```
tests/timeseries/test_quantities.py:47:11: W0143: Comparing against a callable, did you omit the parenthesis? (comparison-with-callable)
```

And `mypy` explodes (again) :boom: (lately I found `mypy` is very unstable, it happens very often that I make a silly change and `mypy` with start throwing an `INTERNAL ERROR`. Sometimes if I run it outside `nox` it will give an error, but sometimes it doesn't. I guess we are taking it to the limit. :unamused: 

Running with `--show-traceback` at least I get which line is making it crash and then looking inside vim, I see the mypy error, so I guess what's killing it is processing all the files at once, and the `mypy` error is:

```
Only instance methods can be decorated with @property  [misc] (mypy error)
```

So weird, because again, Python does support it. OK, researching some more, it seems like it used to work but it was deprecated in 3.11:

> Chaining classmethod descriptors (introduced in bpo-19072) is now deprecated. It can no longer be used to wrap other descriptors such as property. The core design of this feature was flawed and caused a number of downstream problems.

https://github.com/python/mypy/issues/13746#issuecomment-1258864382

Soooooo.... No, we can't do that :P

It would have been nice though, because then it is obvious that it is a singleton.

File: tests/timeseries/test_quantities.py (hunk 1 above)
Comment (shsms) [reply tests/timeseries/test_quantities.py:98]: @llucax we have `+=` now, which mutate the objects.  that's what @Marenz is showing in his example above.

Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: OK, I didn't see the `+=`, that is a problem indeed because it is mutating the object, so yeah, if objects are not immutable it is an issue. I guess we then need to drop the singleton idea, good catch. I guess it is not **that** bad.

Or, drop mutable method. I'm not sure really, there are very nice properties of objects being immutable (like being able to use them as defaults in function arguments). Now I wonder how built-in types work, because they are immutable but `+=` also work on them.

Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: I guess you can just return a new object in `__iadd__`, etc. Should we do that?

Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: It seems like `__i*__` are the only mutating methods. I think for number we really want them to be immutable.

Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: I just tried removing all `__i*__` methods and then `x += ...` with `x = Quantity.zero()` works. I really think we should do that. I think users would expect to be able to use defaults in function arguments, I think `x += ...` should be rare enough to trade a bit of worse performance for safety.

Comment (shsms) [reply tests/timeseries/test_quantities.py:98]: you mean `x += ...` works without `__iadd__`,  and acts as alias for `x = x + ...`?  If that's the case, we can remove that.  Maybe in this PR?

File: src/frequenz/sdk/timeseries/_quantities.py (hunk 2 above)
Comment (Marenz) [reply src/frequenz/sdk/timeseries/_quantities.py:78]: Yeah, I noticed that the mypy people seem to have their own ideas of what is and isn't possible, same with the percentage __mul__ stuff

File: tests/timeseries/test_quantities.py (hunk 1 above)
Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: > you mean x += ... works without __iadd__, and acts as alias for x = x + ...?

Exactly.

>  If that's the case, we can remove that. Maybe in this PR?

I can do that.

File: src/frequenz/sdk/timeseries/_quantities.py (hunk 2 above)
Comment (llucax) [reply src/frequenz/sdk/timeseries/_quantities.py:78]: Well, in this case it is deprecated by Python itself, so it might actually stop working at some point.

File: tests/timeseries/test_quantities.py (hunk 1 above)
Comment (llucax) [reply tests/timeseries/test_quantities.py:98]: Done.

**Comment by llucax** [comment]
Updated to revert adding the mutable methods and added a test to make sure zero is immutable.

**Review by shsms — APPROVED** [review]
